### PR TITLE
New simplify walk

### LINKED ISSUE #410: Why is 'walk' implemented the way it is?
When trying to work around #409, my first thought was to add a walk function for NemoKern that looks in self._kern_schedule instead of self._children, but then I noticed that walk has a very strange implementation:
```
    def walk(self, children, my_type):
        for child in children:
...
            local_list += self.walk(child.children, my_type)
        return local_list
```
Notice that 'self' is only used to determine the walk function used (which as far as I can see is always Node.walk), the parameter children can be something else. That leads to odd calls:
```
nodes = self.root.walk(self.root.children, ...)
```
Why is it not implemented by taking 'children' from self:
```
 def walk(self, my_type):
        for child in self.children:
...
```
then it could be called as
```
nodes = self.root.walk(...)
```
I know there are calls in the form:
```
all_nodes = self.walk(self.root.children, Node)
```
but considering that there appears to be only one walk implementation, I can't see a problem this being changed to:
```
all_nodes = self.root.walk(self.root.children, Node)
```
which would then also allow this to be called as:
```
all_nodes = self.root.walk(Node)
```
with this implementation.

Main reason why I ask: with this implementation I could then overwrite NemoKern.walk to do what I want it to do :to work around #409)

I am happy to discuss this next time we meet, it's not urgent for me.

### PR DESCRIPTION
**Pull request by hiker**
As discussed, this removed the 'children' argument from walk, fixed #410 (and removes a warning or two). This is the rebased version of #426.

### CODE REVIEW

File: src/psyclone/transformations.py [hunk 1]
```diff
@@ -2860,7 +2860,7 @@ def _validate(self, sched):
                 "a schedule of type {0}".format(type(sched)))
 
         # Check that we don't already have a data region of any sort
-        directives = sched.walk(sched.children, Directive)
+        directives = sched.walk(Directive)
         data_directives = [True if isinstance(ddir, (ACCDataDirective,
```
Comment (arporter) [inline src/psyclone/transformations.py]: Again, not yours but pylint says this line can be simplified to something like `data_directives = [bool(isinstance(ddir, (...)) for ddir in directives]`. You might not need the `bool()` constructor though.

File: examples/.gitignore [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+gocean/eg2/inc_field_0_mod.f90
+gocean/eg2/inc_field_1_mod.f90
+gocean/eg3/compute_cu_0.cl
+gocean/eg3/compute_cu_1.cl
+gocean/eg3/compute_cv_0.cl
+gocean/eg3/compute_cv_1.cl
+gocean/eg3/compute_h_0.cl
+gocean/eg3/compute_h_1.cl
+gocean/eg3/compute_z_0.cl
+gocean/eg3/compute_z_1.cl
```
[the remark this answers is not in the document]
Comment (hiker) [reply examples/.gitignore]: Turns out ... you can :)

File: src/psyclone/nemo.py [hunk 3]
```diff
@@ -330,11 +335,17 @@ def match(node):
...
-        if node.walk(node.children, (CodeBlock, NemoLoop)):
-            # A kernel cannot contain unrecognised code (including IO
-            # operations and routine calls) or loops.
-            return False
-        return True
+        # This function is called with node being a NemoLoop. This means
+        # that 'node' is always part of the result of walk. So if there
+        # is a loop or CodeBlock inside, walk will return more than one
+        # element (the first being the node)
+        assert isinstance(node, NemoLoop)
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/nemo.py]: Done, including test case.

File: src/psyclone/psyGen.py [hunk 4]
```diff
@@ -1388,13 +1388,24 @@ def sameParent(self, node_2):
...
 
-    def walk(self, children, my_type):
-        ''' Recurse through tree and return objects of 'my_type'. '''
+    def walk(self, my_type):
+        ''' Recurse through the PSyIR tree and return all objects that are
+        an instance of 'my_type', which is either a single class or a tuple
+        of classes. In the latter case all nodes are returned that are
+        instances of any classes in the tuple.
+
+        :param my_type: The class(es) for which the instances are collected.
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/psyGen.py]: Done.

File: src/psyclone/transformations.py [hunk 5]
```diff
@@ -3081,7 +3081,7 @@ def _validate(self, node_list):
         # Check that we have at least one loop within the proposed region
         found = False
         for node in node_list:
-            loops = node.walk(node.children, Loop)
+            loops = node.walk(Loop)
             if loops or isinstance(node, Loop):
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/transformations.py]: Good catch. I actually changes this code a bit to be more compact (aka pythonic).

File: src/psyclone/psyGen.py [hunk 6]
```diff
@@ -2466,7 +2477,7 @@ def _encloses_omp_directive(self):
             is an error on the part of the user. '''
         # We need to recurse down through all our children and check
         # whether any of them are an OMPDirective.
-        node_list = self.walk(self.children, OMPDirective)
+        node_list = self.walk(OMPDirective)
         if len(node_list) == 0:
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/psyGen.py]: Done.

File: src/psyclone/tests/dynamo0p3_transformations_test.py [hunk 7]
```diff
@@ -6462,12 +6462,12 @@ def test_intergrid_colour_errors(dist_mem, monkeypatch):
     psy = PSyFactory(TEST_API, distributed_memory=dist_mem).create(invoke_info)
     schedule = psy.invokes.invoke_list[0].schedule
     # First two kernels are prolongation, last two are restriction
-    loops = schedule.walk(schedule.children, psyGen.Loop)
+    loops = schedule.walk(psyGen.Loop)
     loop = loops[1]
     # To a prolong kernel
     new_sched, _ = ctrans.apply(loop)
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/dynamo0p3_transformations_test.py]: Done.

File: src/psyclone/transformations.py (hunk 1 above)
Comment (hiker) [reply src/psyclone/transformations.py]: I simplified this a lot further ... using 'any' will allow short-circuit :)

**Comment by hiker** [comment]
Fixed all merging and python2 vs 3 issues, ready for next review @arporter

**Comment by arporter** [comment]
Hi @hiker, did you forget to push your changes - I can't see them?

**Comment by hiker** [comment]
dang ... I think I pushed them to the wrong branch ... will cherry pick the changes across later, apologies.

**Comment by hiker** [comment]
Gee, what a mess - sorry. I hope I cherry picked all the right things now.

File: src/psyclone/transformations.py [hunk 8]
```diff
@@ -3079,13 +3079,11 @@ def _validate(self, node_list):
...
         # Check that we have at least one loop within the proposed region
-        found = False
         for node in node_list:
             loops = node.walk(Loop)
-            if loops or isinstance(node, Loop):
-                found = True
+            if loops:
                 break
-        if not found:
+        else:
```
Comment (arporter) [inline src/psyclone/transformations.py]: Is this correct? If so I don't understand it (although pylint is happy)! I would have expected the `else:` to belong with the `if loops` block. Incidentally I think you could simplify further and just do `if node.walk(Loop):`

File: src/psyclone/transformations.py [hunk 9]
```diff
@@ -2861,10 +2862,9 @@ def _validate(self, sched):
 
         # Check that we don't already have a data region of any sort
         directives = sched.walk(Directive)
-        data_directives = [True if isinstance(ddir, (ACCDataDirective,
-                                                     ACCEnterDataDirective))
-                           else False for ddir in directives]
-        if True in data_directives:
+        if any(isinstance(ddir, (ACCDataDirective,
```
Comment (arporter) [inline src/psyclone/transformations.py]: Ooh, I didn't know about `any`!

File: src/psyclone/tests/nemo/nemo_psy_test.py [hunk 10]
```diff
@@ -274,6 +274,17 @@ def test_codeblock_no_kernel(parser, monkeypatch):
     assert not nemo.NemoKern.match(loop)
 
 
+def test_match():
```
Comment (arporter) [inline src/psyclone/tests/nemo/nemo_psy_test.py]: Please rename this to `test_kernel_match` or somesuch.

File: src/psyclone/psyGen.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyGen.py]: Thanks.

**Pull request merged**